demo/full/scripts/modules/bxf.js

```diff
@@ -0,0 +1,673 @@
+};
+const channelLogoURL = "https://drm.canal-plus.com:8443/bxf-parser2/tmp/series.png";
+const timeShiftBufferDepth = 60 * 60 * 8;
+
+function getURLForPicto(title) {
+  return pictos[title];
+}
+
+function getLicenseURL(initData) {
+  const keyid = parseInitData(initData).toUpperCase();
+  if (!keyid) {
+    throw "ERROR";
+  }
+  return kidByAffairePgrm[keyid];
+}
+
+function getLicense(challenge, _, initData) {
+  const licenseURL = getLicenseURL(initData);
+  return new Promise((resolve) => {
+    getLicenseObject(token, licenseURL, challenge).then((license) => {
+      resolve(license);
+    });
+  });
+}
+
+function getServerCertifcate() {
+  const xhr = new XMLHttpRequest();
+  xhr.open("GET", "https://secure-webtv-static.canal-plus.com/widevine/cert/cert_license_widevine_com.bin", true);
+  xhr.responseType = "arraybuffer";
+  return new Promise((resolve) => {
+    xhr.onload = (evt) => {
+      resolve(evt.target.response);
+    };
+    xhr.send();
+  });
+}
+
+function parseInitData(initData) {
+  const textDecoder = new TextDecoder("utf-8");
+  function getAllIndexes(arr, val) {
+    const indexes = [];
+    let i = -1;
+    while ((i = arr.indexOf(val, i+1)) != -1) {
+      indexes.push(i);
+    }
+    return indexes;
+  }
+  const idx = getAllIndexes(textDecoder.decode(initData), "pssh");
+  const psshs = idx.reduce((acc, value) => {
+    const begin = value - 4;
+    let position = begin;
+    position += 8;
+    const version = be4toi(initData, position);
+    position += 4;
+    if (version === 0) {
+      position += 16;
+      const privateLength = be4toi(initData, position);
+      position += 4;
+      const privateData = initData.subarray(position, position + privateLength);
+      acc.push(privateData);
+    }
+    return acc;
+  }, []);
+  const widevinePssh = psshs.find((pssh) => pssh.length === 20);
+  if (widevinePssh) {
+    return bytesToHex(widevinePssh.subarray(4, 20), "-");
+  }
+  return undefined;
+}
+
+export default function loadBXF({
+  url: bxfURL,
+  textTrackElement,
+  overlayElement,
+  textTrackMode,
+  beginning,
+}) {
+  return parseBXF(bxfURL, BXFParser, "text").then(async ({ contents: parsedContents }) => {
+    /* eslint-disable no-console */
+    console.log("############## Parsed BXF ################");
+    /* eslint-enable no-console */
+
+    getKID(fillingManifestURL).then(({ kid }) => {
+      kidByAffairePgrm[kid] = fillingLicenseURL;
+    });
+
+    const epg = [];
+    const contents = [];
+    const overlays = [];
+    const logoOverlays = [];
+
+    const promises = [];
+
+    for (const parsedContent of parsedContents) {
+      promises.push(
+        new Promise((resolve, reject) => {
+          const videos = parsedContent.video;
+          if (videos) {
+            for(const video of videos) {
+              getManifestURL(token, video.affaire, video.pgrm)
+                .then(({ manifestURL, licenseURL }) => {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Define this function outside of a loop. ([javascript:FunctionDefinitionInsideLoop](https://sonarcloud.io/organizations/rx-player/rules#rule_key=javascript%3AFunctionDefinitionInsideLoop))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=366&issues=AWbt7Z5cW3Y69bz5aGXS&open=AWbt7Z5cW3Y69bz5aGXS)

[ISSUE_KEY]: AWbt7Z5cW3Y69bz5aGXS